In [1]:
# Import packages

import pandas as pd
import numpy as np 

In [2]:
df = pd.read_excel("C:\Portfolio\Data Engineer Track\Airline Datasets and Courses\PWC_The Forage Power BI Course\Task 2_Customer Retention\Churn-Dataset.xlsx")

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [4]:
df.shape

(7043, 23)

In [5]:
df = df.drop(["customerID","gender", "Partner",	"Dependents","PhoneService","MultipleLines","InternetService","OnlineSecurity","OnlineBackup","DeviceProtection","TechSupport","StreamingTV","StreamingMovies","Contract","PaperlessBilling","PaymentMethod"], axis = 1)



In [6]:
one_hot_encoded_df = pd.get_dummies(df, columns = ["Churn"])
print(one_hot_encoded_df)

      SeniorCitizen  tenure  MonthlyCharges  TotalCharges  numAdminTickets  \
0                 0       1           29.85         29.85                0   
1                 0      34           56.95       1889.50                0   
2                 0       2           53.85        108.15                0   
3                 0      45           42.30       1840.75                0   
4                 0       2           70.70        151.65                0   
...             ...     ...             ...           ...              ...   
7038              0      24           84.80       1990.50                0   
7039              0      72          103.20       7362.90                0   
7040              0      11           29.60        346.45                0   
7041              1       4           74.40        306.60                0   
7042              0      66          105.65       6844.50                2   

      numTechTickets  Churn_No  Churn_Yes  
0                  

In [7]:
one_hot_encoded_df = one_hot_encoded_df.drop(["Churn_Yes"], axis = 1)

In [8]:
one_hot_encoded_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   SeniorCitizen    7043 non-null   int64  
 1   tenure           7043 non-null   int64  
 2   MonthlyCharges   7043 non-null   float64
 3   TotalCharges     7043 non-null   float64
 4   numAdminTickets  7043 non-null   int64  
 5   numTechTickets   7043 non-null   int64  
 6   Churn_No         7043 non-null   bool   
dtypes: bool(1), float64(2), int64(4)
memory usage: 337.1 KB


In [9]:
one_hot_encoded_df = one_hot_encoded_df.apply(pd.to_numeric)

In [10]:
one_hot_encoded_df.corr()

SeniorCitizen    tenure  MonthlyCharges  TotalCharges  \
SeniorCitizen         1.000000  0.016567        0.220173      0.102652   
tenure                0.016567  1.000000        0.247900      0.825464   
MonthlyCharges        0.220173  0.247900        1.000000      0.650864   
TotalCharges          0.102652  0.825464        0.650864      1.000000   
numAdminTickets      -0.000881 -0.000440        0.007386      0.006953   
numTechTickets        0.093330  0.215158        0.245081      0.283985   
Churn_No             -0.150889  0.352229       -0.193356      0.199037   

                 numAdminTickets  numTechTickets  Churn_No  
SeniorCitizen          -0.000881        0.093330 -0.150889  
tenure                 -0.000440        0.215158  0.352229  
MonthlyCharges          0.007386        0.245081 -0.193356  
TotalCharges            0.006953        0.283985  0.199037  
numAdminTickets         1.000000       -0.008004  0.019877  
numTechTickets         -0.008004        1.000000 -0.357283  
Churn_No                0.019877       -0.357283  1.000000

In [11]:
X = one_hot_encoded_df.drop(["Churn_No"], axis = 1)
y = one_hot_encoded_df["Churn_No"]

In [12]:
# Import data science and machine learning packages

from sklearn.model_selection import train_test_split
from sklearn import preprocessing
# import module
from sklearn.preprocessing import MinMaxScaler

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.17, random_state = 123)

# Scale the features using minmaxscaler
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [13]:
# Import Support Vector Classifier packages, fit data, and print model accuracy

from sklearn.svm import SVC 

svm = SVC(random_state = 123)

svm.fit(X_train, y_train)

print("print accuracy of svm model:", svm.score(X_test, y_test))

print accuracy of svm model: 0.8514190317195326


In [14]:
# Import necessary modules for K-Nearest Neighbors

from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=3)
  
knn.fit(X_train, y_train)
  
# Calculate the accuracy of the model
print(knn.score(X_test, y_test))

0.832220367278798


c:\Users\rakim\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


The most accuracy model is the support vector machine classifier (85%). 